src/shogun/base/SGObject.h

```diff
@@ -113,6 +113,32 @@ using stringToEnumMapType = std::unordered_map<std::string, std::unordered_map<s
 			this->m_gradient_parameters->add(param, name, description);        \
 	}
 
+#define SG_ADD_CONSTRAINT(param, name, description, param_properties, ...)     \
+	{                                                                          \
+		static_assert(                                                         \
+		    static_cast<bool>(                                                 \
+		        (param_properties)&ParameterProperties::CONSTRAIN),            \
+		    "Expected param to have ParameterProperty::CONSTRAIN");            \
+		AnyParameterProperties pprop =                                         \
+		    AnyParameterProperties(description, param_properties);             \
+		this->m_parameters->add(param, name, description);                     \
+		this->watch_param(                                                     \
+		    name, param,                                                       \
+		    [](auto val) {                                                     \
+			    bool state;                                                    \
+			    auto lambda = [&state](auto casted_val) {                      \
+				    state = make_composer(__VA_ARGS__).run(casted_val);        \
+			    };                                                             \
+			    sg_any_dispatch(val, sg_all_typemap, lambda);                  \
+			    return state;                                                  \
+		    },                                                                 \
```

Comment:
this is the best I could come up with without using a stateful lambda (which can't be casted to a function pointer)

Reply:
state stores the result of the evaluation of the constraint functions